### ISSUE TITLE
Make camera raw preview available to compare with current edit

### ISSUE THREAD

**Issue by Waiwei8k** [darktable-org/darktable#5778]
Camera raw preview images can be a usefulhint for raw processing. Often they contain sensible colour rendering defaults and are at least a consistent reminder of the scene as shot.

I would like a way to conveniently compare my current processed image to the RAW file's embedded jpg.

This might be a pre-populated snap-shot in the snapshot dialogue (my favourite option), a module which would introduce the embedded JPG into the pixel pipe with all the usual opacity/masking options, an entry in the duplicate manager or whatever best fits the code.

**Comment by parafin** [comment]
Not sure it's a good idea for general public - there's already a lot of questions "how to get the same result as camera JPEG" even without proposed easier access to it...

**Comment by Waiwei8k** [comment]
I don't think this will make things any worse, but certainly anyone who does want to match camera JPEG results can use this feature as a starting point. It may be a good way to keep those people happy whilst encouraging them to learn how to drive the tool. Conversely, it's often quite nice to see how much you've improved the picture over the camera's attempt.

**Comment by wpferguson** [comment]
There is a lua-script, examples/multi_os, that does this.  The embedded jpg is extracted, imported, and grouped with the raw for each selected image.

**Comment by elstoc** [comment]
I like this idea but I wouldn't want it as a separate image in my library (grouped or otherwise). The embedded JPEG is already displayed in the lighttable before you load the image in the darkroom so people can already view it - it's just that the only way to _keep_ it available (in the lighttable only) is to have a separate duplicate version with no edits.

**Comment by wpferguson** [comment]
> The embedded JPEG is already displayed in the lighttable

Only if you leave the preview as the default.  If you select half size raw, then this isn't the case.

**Comment by elstoc** [comment]
Or it could be an extra option in the duplicate module and you could do the same.

**Comment by wpferguson** [comment]
The duplicate is a duplicate of the raw image.  The embedded jpg is a different image, extracted from the raw, so it's not actually a duplicate.

**Comment by elstoc** [comment]
It could have its own independent keyboard shortcut and/or button on the bottom of the darkroom then

**Comment by aurelienpierre** [comment]
Having that in snapshot or duplicate module is not super difficult to do, however, what should we do with cropping/rotation/perspective distortions ? 

1. Apply the same distortions on the JPEG thumbnail ? 
2. Keep it as-is ?

**Comment by elstoc** [comment]
I'd say we should keep the JPEG unmodified - treat it as we would a "differently-edited" duplicate in the duplicate module. For example, the duplicate module should be able to handle images that have different crops and/or where one image has lens correction enabled and one does not. We shouldn't worry about those inconsistencies.

**Comment by Waiwei8k** [comment]
The JPG may well already have a lens correction applied in camera, so no need to apply that. An option to match the crop would be nice, but as a very low priority addition. Certainly not required.

**Comment by Waiwei8k** [comment]
Pinging to stop auto-close. I would still like this feature.

**Comment by johnny-bit** [comment]
I think the best option here would be to use snapshot module for this and on opening image create snapshot from extracted JPEG preview from RAW file. Don't do ANY processing on it. Just thumbnail dump onto snapshot. And hide that behind some option. Would be usable enough IMO.

**Comment by Waiwei8k** [comment]
Just pinging because I don't want this to time out.

**Comment by Waiwei8k** [comment]
Another ping to keep this live.

**Comment by afranke** [comment]
> I think the best option here would be to use snapshot module for this and on opening image create snapshot from extracted JPEG preview from RAW file. Don't do ANY processing on it. Just thumbnail dump onto snapshot. And hide that behind some option. Would be usable enough IMO

That sounds good and unobtrusive. In addition to that, it could be handy to have a side by side mode similar to culling in the lighttable, but just the snapshot idea would already be very useful.

**Comment by Waiwei8k** [comment]
Pinging to stop auto-close. I would still like this feature.

**Comment by ptilopteri** [comment]
you already have the requested ability.

duplicate your image
remove all edits from the dup
view original and dup to compare

**Comment by Waiwei8k** [comment]
Do you mean duplicate and remove history stack? When you go into the darkroom view, the default stack gets processed and you get the RAW again, so there's no way to put the image on the snapshot viewer stack. Sorry if I've miss-understood your suggestion.

**Comment by ptilopteri** [comment]
* Waiwei8k ***@***.***> [06-02-24 18:09]:
> Do you mean duplicate and remove history stack? When you go into the darkroom view, the default stack gets processed and you get the RAW again, so there's no way to put the image on the snapshot viewer stack. Sorry if I've miss-understood your suggestion.

> Message ID: ***@***.***>

open your image in dt
take a snapshot
open history panel
select "O original" and take a snapshot

and you have both images

or I have no idea your purpose.

-- 
(paka)Patrick Shanahan       Plainfield, Indiana, USA          @ptilopteri
facebook/ptilopteri
Photos: http://wahoo.no-ip.org/piwigo                   paka @ IRCnet oftc

**Comment by afranke** [comment]
> you already have the requested ability.

No. We are asking for the embedded jpeg, which is processed by the camera and stored in the RAW file, not the bare RAW with no edits as you are suggesting. It is shown in the lighttable before any edits are made, and then it is not available anymore. This issue is asking for a way to access it from the darkroom at any given moment, to compare with current edit, as the title says.

**Comment by wpferguson** [comment]
There is a lua-script that extracts the embedded jpg and groups it with the raw in lighttable, `examples/multi_os`.

The extracted jpg is not the same resolution as the raw, at least in recent cameras, so overlaying won't work.

**Comment by afranke** [comment]
I understand this may not be the case for all formats, but my camera (Canon EOS R6) has a full size JPG available (`JpgFromRaw` for exiftool). We are also not after extreme precision here. The point is to use it as a reference, not to match details.

**Comment by wpferguson** [comment]
So why not just shoot raw+jpeg?

**Comment by Waiwei8k** [comment]
Jump though enough hoops and you can do anything, but that misses the point. The real value is having the image easily available with minimum friction. Also, different resolutions could be resolved by scaling the embedded image.

**Comment by victoryforce** [comment]
> The real value is having the image easily available with minimum friction

This will be done soon with the addition of system styles that replicate the processing of specific cameras.

**Comment by Waiwei8k** [comment]
If it respects the processing settings you have on the camera, that would be a reasonable alternative and a great addition. Sounds like a huge amount of work.

**Comment by Waiwei8k** [comment]
Ping to mark as not stale.

**Comment by Waiwei8k** [comment]
Ping to mark as not stale.

**Comment by Waiwei8k** [comment]
Ping to mark as not stale.

**Comment by sebkaletta** [comment]
I like this idea.
It would be possible to check if your edits beat the OOC jpegs. Especially if you're a beginner, it's great to see if you didn't overcorrect something or completely went the wrong direction. I think that would boost the learning curve, as you can directly have some sort of feedback, and not wait to the next day, thinking e.g. "why did I pull the saturation this high".

I think for now even the possibility to open the preview jpeg in a dropdown menu on the sidebar would be sufficient.

**Comment by Waiwei8k** [comment]
Ping to mark as not stale.

**Comment by Waiwei8k** [comment]
Ping to mark as not stale.

**Comment by Waiwei8k** [comment]
Ping to mark as not stale.

**Comment by Waiwei8k** [comment]
Ping to mark as not stale.

**Comment by Waiwei8k** [comment]
Ping to mark as not stale.

**Comment by afranke** [comment]
🤸‍♂️

**Comment by wpferguson** [comment]
@TurboGit (I'm pinging you because IIRC you did the current snapshot module) @jenshannoschwalm (because you're the mipmap god) would it be possible to generate a snapshot from the large (size 10) jpeg mipmap?  

IIUC the reason is more for look so I don't think the snapshot would have to match the "distortions" made to the image.

**Comment by TurboGit** [comment]
@wpferguson : No not at the moment. We generate a preview based on the current edit stack. That is, the RAW (or JPEG) and a specific step in the history range. Having a JPEG coming from another source is not current supported and so not possible. Given that this is a really specific scenario I don't think it will be implement in core darktable.

The current solution is to extract the JPEG (or shot RAW+JPEG), to import it. You can then snapshot it, move to the RAW and compare.

**Comment by wpferguson** [comment]
> We generate a preview based on the current edit stack.

That's what I thought.

> The current solution is to extract the JPEG (or shot RAW+JPEG), to import it. You can then snapshot it, move to the RAW and compare

Maybe I can automate that somehow...

**Comment by outfrost** [comment]
Since the goal of having easy access to "peek" the ooc JPEG without polluting the library with JPEGs for every raw photo doesn't seem compatible with the design of either the snapshot module, nor the duplicate system, perhaps a separate darkroom module could do the trick? I imagine it could let the user choose whether to use raw-embedded JPEG or a sibling image file (depending on which the camera saved and at what resolution), and press a button to toggle a comparison similar to what snapshot does. Applying crop and rotation would be very useful, lens correction should probably stay optional as others have noted.

**Comment by wpferguson** [comment]
This is a custom workflow.  More than that it's a "housekeeping" request and not a processing (image development) request.  The whole point is that the user wants to have the usage of the embedded/paired jpeg without having the actual image file, kind of like I want to have my cake and eat it too :-).

Lua scripting is the tool for implementing custom workflows.  So, we could have a script that:

- creates a module in lighttable/darkroom similar to the duplicate module 
- when invoked it creates a link (or copies the file on windows) to the full resolution mipmap
- imports the linked/copied image and groups it with the original
- removes the linked/copied jpeg when:
  * invoked from the module
  * darktable exits
  * the view changes to lighttable

**Comment by jenshannoschwalm** [comment]
> kind of like I want to have my cake and eat it too :-).

Don't understand this :-) I would have thought the equivalent would be "i don't want the cake but want to eat it" :-) 

Whatever - otherwise absolutely agreed.
Mostly a typical "new user desire" but this feature might also be used to trim camera-specific presets.

**Comment by Waiwei8k** [comment]
I don't get this odd obsession with any reference to JPEGs being a "new user" thing. I word it like that because I see similar sentiment across many forums. I'm far from a new user and don't want to copy the JPEG as a rule. Yes, I know there are some new users who consider "correct" to be what the JPEG does, but that's far from the only use case for referring to the camera output. Sometimes the camera default does some interesting things and sometimes a specific aspect of that is worth investigating alongside my darktable edit.

As to how this request be best implemented, obviously devs will have to make the final call, but I hope that the best user interface is given higher priority than the easiest implementation. If no one wants to do the work, totally fair, I express what I would like to see and don't pretend to have any entitlement where other people's work is involved.

**Comment by jenshannoschwalm** [comment]
> I don't get this odd obsession with any reference to JPEGs being a "new user" thing.

Well - it's neither "odd" nor an "obsession". It's just a clear observation.

This discussion is **very** long and old, it's pretty clear that none of the devs is interested in implementing this as a core feature.

Doing it via lua as a "workflow feature support" would be a way.

Just shooting JPEG+RAW would be the recommended way to "compare". Adding the "desired feature" would increase code & UI complexity a lot, would always miss some features, would be an absolute use-cornercase and as saying so would simply be a waste of time & resources. How to support all that modern camera LUT stuff? Or distorting stuff offered by some cameras? Film simulations? **IF** you want to play around, go for it with jpgs and compare vs what you do in dt with the raw file.

So in short - i don't see any chance to get this implemented in dt and i even would definitely not want it as a core feature.
dt is mainly a raw processor, that's the way the cookie crumbles.

**Comment by Waiwei8k** [comment]
Could you please elaborate on the need to support LUTs for this feature? I was imagining just adding the JPEG as a preset (or fetch button) to the snapshot module. The JPEG should be precomputed by the camera. The largest issue I can see is if the JPEG has a difficult colour space?

Just to check that you don't think I'm asking for a JPEG look to be auto applied to the processing pipeline as that is not my ask.

**Comment by jenshannoschwalm** [comment]
> Could you please elaborate on the need to support LUTs for this feature? I was imagining just adding the JPEG as a preset (or fetch button) to the snapshot module. The JPEG should be precomputed by the camera. The largest issue I can see is if the JPEG has a difficult colour space?

I didn't say at all there is any need to support LUT's for this feature. I just was hinting that modern cameras offer a huge amount of presets/tunings .. to modify the jpeg output. Think about fujis or pana ... Embedded jpegs - what do they contain? Just a preview, nothing more. What is presented there is absolutely free to the vendor. Some offer good quality images, some only little thumbs, ... And for sure, the idea that the embedded jpeg shows the scene as shot is just plain wrong.

One more example. Some cameras offer cropping. Raw image provides full sensor data nevertheless in many cases, the embedded jpeg shows the crop. And there are many, many more little peculiar things. 

So - this FR request has been opened some years ago and none of the devs has ever been interest to get into this. Why? None of the devs sees this feature of any importance worth to spend a lot of time/effort or some devs even think "i donj't want it". Or might just think "how boring". 

Sure - of course you can ping this feature request to stay open as often as you like. But - frankly speaking - none of the folks reading/commenting/downthumbing here should expect this to be implemented in any means. That's simply not as development works. All devs are simply absolutely free to chose what they do (and we listen to user requests/wishes a lot and do things we personally wouldn't even use). Or - you should just respect what other people are doing and what they are not doing!

As always - that's how i got into dt development - if there is a missing feature that itching you and no other person seems to be interested, you have to get into coding yourself.

**Comment by Waiwei8k** [comment]
>  And for sure, the idea that the embedded jpeg shows the scene as shot is just plain wrong.

Obviously. Why bring up that straw-man here? Hinting that this request is anything to-do with that misunderstanding is not useful.

> One more example. Some cameras offer cropping. Raw image provides full sensor data nevertheless in many cases, the embedded jpeg shows the crop. And there are many, many more little peculiar things.

Yes, dealing with that sort of thing is a nice to have, but not required for the feature to be useful.

> Sure - of course you can ping this feature request to stay open as often as you like. But - frankly speaking - none of the folks reading/commenting/downthumbing here should expect this to be implemented in any means. That's simply not as development works. All devs are simply absolutely free to chose what they do (and we listen to user requests/wishes a lot and do things we personally wouldn't even use). Or - you should just respect what other people are doing and what they are not doing!

I covered that explicitly in my previous post. It feels like you are personally offended by this request. I'm happy to keep the request open in the hope that a dev sees the value and takes it on. If they don't, totally fine. I don't expect anything of them as I said before. Also, the downthumb wasn't from me.

> As always - that's how i got into dt development - if there is a missing feature that itching you and no other person seems to be interested, you have to get into coding yourself.

Perhaps. I might be able to find time in a few years to learn some of this stuff.

Edit: Fixed formatting.

**Comment by outfrost** [comment]
@wpferguson 
> This is a custom workflow.

Isn't the entirety of darktable about custom workflows? Why bother with all the sliders, curves and custom styles otherwise?

> kind of like I want to have my cake and eat it too :-)

There are less condescending ways to say that you don't like the idea and won't allow it into dt core.

I could point at the history, snapshot and duplicate modules, and say that you can _technically_ accomplish what each of them does using another one and a bit of manual labour (or lua scripting). Yet obviously nobody's rushing to deprecate or remove any of them, and I make frequent use of all 3, because they excel at different tasks.

**Comment by jenshannoschwalm** [comment]
> I could point at the history, snapshot and duplicate modules, and say that you can _technically_ accomplish what each of them does using another one and a bit of manual labour (or lua scripting). 

That's just absolute crap / trolling.

**Comment by afranke** [comment]
> One more example. Some cameras offer cropping. Raw image provides full sensor data nevertheless in many cases, the embedded jpeg shows the crop. And there are many, many more little peculiar things.

Which is a good reason to show the JPEG preview. You may decide on some settings on your camera as you shoot to have an idea of the final result in camera, but use RAW for the freedom it gives you in post processing. Being able to see those settings in the darkroom gives some insight into what you had in mind during the shooting session. “Oh, I remember now that I thought this one would work well in black & white, in a square ratio”.

Shooting RAW+JPG results in an unnecessary overhead.

**Comment by jenshannoschwalm** [comment]
@afranke so - as said - if it's itching you, go for it.

**Comment by wpferguson** [comment]
> Isn't the entirety of darktable about custom workflows?

No, it's about workflows.  darktable provides tools and users arrange them in a sequence to accomplish a task.  For instance I could assign shortcuts to 10 different modules to apply each with a designated preset.  That's my workflow.  But then I decide I'm getting carpal tunnel from using 10 shortcuts on every image, so I want to consolidate that into just one shortcut.  darktable doesn't support that out of the box, so now I need a custom work flow to accomplish it.  So I create a script that exposes one shortcut and when I trigger that shortcut the script triggers the other 10 shortcuts.  

Custom workflows are workflows that darktable can't do without being extended in some manner.

> There are less condescending ways to say that you don't like the idea and won't allow it into dt core.

The description was accurate.  This isn't about adding functionality to darktable.  darktable already does what the users want from a JPEG image, it just requires the image to do it.  This is simply about not having the file on disk. So to implement it in core:

- we have to extract the preview from the RAW.  Not all images contain full size previews.
- we have to keep the preview in memory which impacts the amount of memory available for processing the image thus slowing down darkroom
- we have to rewire all the functions that work on a disk image to work on a memory image
- we have to do more processing which slows darktable down
- it adds a lot of code and work that provides **no functionality** but has to be maintained

---

> Shooting RAW+JPG results in an unnecessary overhead.

Some cameras don't have full size previews, so shooting RAW+JPG is necessary if you want the full size preview.

**Comment by Waiwei8k** [comment]
> * it adds a lot of code and work that provides **no functionality** but has to be maintained

It provides a better user experience taking the feature from too clumsy to be useful to genuinely useful.

If a function is not suitably usable it might as well not exist.

**Comment by ptilopteri** [comment]
As many apps are available to provide the embedded jpg image, why not use one of them rather than increasing the complexity of dt and the work load upon the developers leaving time to work on actual problems?

**Comment by Waiwei8k** [comment]
> leaving time to work on actual problems?

I clearly disagree with your framing. The request would improve dt.

> As many apps are available to provide the embedded jpg image, why not use one of them

I'm not trying to just see the embedded jpg. I want to have it in the snapshot window for easy comparison my edit.

I get that there are plenty of people who don't see the point of this and don't feel it's worth their time and effort to implement it. No one is expecting that they should.

**Comment by outfrost** [comment]
@wpferguson 
> This is simply about not having the file on disk.

Since when exactly? There must've been a misunderstanding about not having the file in dt library vs not having the file on disk.

> So I create a script that exposes one shortcut and when I trigger that shortcut the script triggers the other 10 shortcuts.

In your described use case the scripting solution ends up having perfectly acceptable UX. That's the difference.

> Custom workflows are workflows that darktable can't do without being extended in some manner.

In other words, anything not already implemented in core is considered "custom"? Are any new UX features therefore banned? Stating that early on could've avoided a lot of confusion, and basically this entire long issue thread.

@jenshannoschwalm 
> That's just absolute crap / trolling.

That's so kind and insightful of you!

I'm going to refrain from interacting here from now on, since user feedback and discussion on possible solutions are seen by maintainers as invalid and hostile.

**Comment by wpferguson** [comment]
@outfrost did you want to argue (which is how the "tone" of the post sounds to me) or were you asking questions?

> Since when exactly? There must've been a misunderstanding about not having the file in dt library vs not having the file on disk.

From OP: This might be a pre-populated snap-shot in the snapshot dialogue (my favourite option), a module which would introduce the embedded JPG into the pixel pipe with all the usual opacity/masking options, an entry in the duplicate manager or whatever best fits the code.

another comment above: I wouldn't want it as a separate image in my library (grouped or otherwise). 

The simple fact is that shooting RAW+JPEG lets you do what OP asked, have a snapshot of the JPG available to overlay the raw.

It's possible to have an image without a file on disk, but it looks like this - 

<img width="107" height="143" alt="Image" src="https://github.com/user-attachments/assets/344dcca0-61b8-4062-a94e-682d276d9378" />

and you can't edit it or take a snapshot of it.

---

> In other words, anything not already implemented in core is considered "custom"? Are any new UX features therefore banned?

Not what I said and not what my example showed. Lua can do things, including implement new interfaces, without any change to master.  darktable provides a set of "user ready" tools to do image processing in a workflow of their choosing.  Lua provides a set of "some assembly required" tools to do things that allow more flexibility for workflows (custom) that don't fit the "user ready" tools.  

Changes to master implement new tools (AI), or fix broken ones that can be incorporated into user workflows.  If there is a way to accomplish the workflow, which in this case there is (shoot RAW+JPEG and take a snapshot of the JPEG) then changing master isn't required since master provides the tools to accomplish the task.  If the user doesn't like the workflow (having to shoot RAW+JPEG) then we resort to the "some assembly required" tools to try and provide a more suitable (custom) workflow.

---

> hostile

That's definitely how I interpreted this post.  My first reaction was to close the issue and be done with it.  Then I walked away for a couple of hours.

I've assigned this to myself and taken it on, so I'm not sure why I'm getting yelled at for trying to help.

**Comment by Waiwei8k** [comment]
Another ping to keep this live.
